File: algorithms/sorting/selection_sort/ruby/selection_sort.rb
def selection_sort(array)
  sorted_array = []

  array.size.times do
    smallest_index = find_smallest_element_index(array)
    sorted_array.append(array.delete_at(smallest_index))
  end

  sorted_array
end


def find_smallest_element_index(array)
  return if array.empty?

  smallest = array.first
  smallest_index = 0

  array.each_with_index do |element, index|
    if element < smallest
      smallest = element
      smallest_index = index
    end
  end

  smallest_index
end

File: algorithms/sorting/selection_sort/javascript/selectionSort.js
const selectionSort = (array) => {
  let sortedArray = [];
  let iterations = array.length

  for (let i = 0; i < iterations; i++) {
    let smallestIndex = findSmallestElementIndex(array);
    let smallest = array[smallestIndex];

    array = array.filter(e => e !== smallest);
    sortedArray.push(smallest);
  }

  return sortedArray;
};


const findSmallestElementIndex = (array) => {
  if (array.length === 0) return null ;

  let smallestIndex = 0;
  let smallest = array[smallestIndex];

  for (let i = 1; i < array.length; i++ ) {
    if (array[i] < smallest) {
      smallestIndex = i;
      smallest = array[i];
    }
  }

  return smallestIndex;
};

module.exports = { findSmallestElementIndex, selectionSort };
